packages/@lwc/template-compiler/src/shared/ir.ts

```diff
@@ -42,105 +141,293 @@ export function createElement(original: parse5.Element): IRElement {
-    }
+export function ifBlock(
+    name: string,
+    location: SourceLocation,
+    modifier: string,
+    condition: Expression
+): IfBlock {
+    return {
+        type: LWCNodeType.IfBlock,
+        name,
+        location,
+        children: [],
+        modifier,
+        condition,
+    };
+}
 
+export function eventListener(
+    name: string,
+    handler: Expression,
+    location: SourceLocation
+): EventListener {
     return {
-        type: 'text',
+        type: LWCNodeType.EventListener,
+        name,
+        handler,
+        location,
+    };
+}
+
+export function keyDirective(value: Expression, location: SourceLocation): KeyDirective {
+    return {
+        type: LWCNodeType.Directive,
+        name: LWCNodeType.Key,
         value,
-        location: original.sourceCodeLocation,
+        location,
     };
 }
 
-export function createComment(original: parse5.CommentNode, value: string): IRComment {
-    if (!original.sourceCodeLocation) {
-        throw new Error('Invalid comment AST node. Missing source code location.');
-    }
+export function dynamicDirective(value: Expression, location: SourceLocation): DynamicDirective {
+    return {
+        name: LWCNodeType.Dynamic,
+        value,
+        type: LWCNodeType.Directive,
+        location,
+    };
+}
+
+export function domDirective(lwcDomAttr: string, location: SourceLocation): DomDirective {
+    return {
+        type: LWCNodeType.Directive,
+        name: LWCNodeType.Dom,
+        value: {
+            type: LWCNodeType.Literal,
+            value: lwcDomAttr as LWCDirectiveDomMode,
+        },
+        location,
+    };
+}
 
+export function preserveCommentsDirective(
+    preserveComments: boolean,
+    location: SourceLocation
+): PreserveCommentsDirective {
     return {
-        type: 'comment',
+        name: LWCNodeType.PreserveComments,
+        value: {
+            type: LWCNodeType.Literal,
+            value: preserveComments,
+        },
+        type: LWCNodeType.Directive,
+        location,
+    };
+}
+
+export function renderModeDirective(
+    renderMode: string,
+    location: SourceLocation
+): RenderModeDirective {
+    return {
+        name: LWCNodeType.RenderMode,
+        value: {
+            type: LWCNodeType.Literal,
+            value: renderMode as LWCDirectiveRenderMode,
+        },
+        type: LWCNodeType.Directive,
+        location,
+    };
+}
+
+export function attribute(
+    name: string,
+    value: Expression | Literal,
+    location: SourceLocation
+): Attribute {
+    return {
+        type: LWCNodeType.Attribute,
+        name,
         value,
-        location: original.sourceCodeLocation,
+        location,
+    };
+}
+
+export function property(
+    name: string,
+    value: Expression | Literal,
+    location: SourceLocation
+): Property {
+    return {
+        type: LWCNodeType.Property,
+        name: attributeToPropertyName(name),
```

Comment:
Those helper methods should be pretty dumb. The attribute to prop name conversion should be done by the invoker. And fortunately, the is only a single invoker for this method.